### DISCUSSION TITLE
atomWithStorage without initial value

### DISCUSSION THREAD

**Discussion by mwisnicki** [pmndrs/jotai#1784] [Ideas]
I would like to have `atomWithStorage` that blocks until value is read from `AsyncStorage` so that all consumers are blocked until it is resolved.

Currently this is impossible because initialValue is required and also `atomWithStorage<Promise<any>>`  expects storage of type `AsyncStorage<Promise<any>>`.

**Comment by dai-shi** [comment]
Thanks for the suggestion. I understand the requirement, but it's out of the scope of current `atomWithStorage` which is more focused on SSR friendliness.

I think the best is to create a `atomWithAsyncStorage` recipe in docs.
`atomWithAsyncStorage` should be implemented with `atomWithObservable`.
Would anyone like to give a try?

**Reply by mwisnicki** [reply]
This is what I was able to come up with:

https://codesandbox.io/s/jotai-atomwithstorageasync-qjl94x?file=/src/utils.ts

**Reply by mwisnicki** [reply]
When I tried to defer loading until mount I've run into an issue where atoms are not mounted until initial promise is resolved:

https://codesandbox.io/s/jotai-atomwithstorageasync-forked-8br4yb?file=/src/utils.ts

How else can I create blocking atom other than `atom(new Promise(() => {}))`?

**Reply by dai-shi** [reply]
I really suggest to build on top of `atomWithObservable`. That's a battle tested util for such purpose.

**Reply by mwisnicki** [reply]
What do you think about this: https://codesandbox.io/s/jotai-atomwithstorageasync-atomwithobservable-2ngl64?file=/src/utils.ts

**Reply by dai-shi** [reply]
Hmm, I wonder if we can simply create an observer (a subject) from AsyncStorage. Then, it's something like the following.

```ts
export function atomWithStorageAsync<Value>(
  key: string,
  storage: AsyncStorage<Value>
): WritableAtom<
  Maybe<Value> | Promise<Maybe<Value>>, // XXX it's always promise
  [SetStateActionWithoutInitial<Value>],
  Promise<void>
> {
  const subject = createObserverFromAsyncStorage(storage);

  const baseAtom = atomWithObservable(subject);

  return atom(
    (get) => get(baseAtom),
    async (get, set, update: SetStateActionWithoutInitial<Value>) => {
      const nextValue =
        update instanceof Function ? update(await get(baseAtom)) : update;
      set(baseAtom, nextValue);
    }
  );
}
```

**Reply by mwisnicki** [reply]
It can be done although without rxjs it's even longer:
https://codesandbox.io/s/jotai-atomwithstorageasync-atomwithobservable-v2-lov3ck?file=/src/utils.ts

**Reply by dai-shi** [reply]
looks great to me.

**Reply by mwisnicki** [reply]
Do you know if I need to unsubscribe from storage in complete handler?

**Reply by dai-shi** [reply]
tbh, I don't know. With `atomWithObservable` usage, it never completes anyway.

**Reply by jetaggart** [reply]
Will this be included in utils at some point? This is a very useful atom

**Reply by dai-shi** [reply]
Basically, we wouldn't consider adding derived utils in jotai/utils. Our options are, a) recipes in docs and/or b) jotaijs/* packages.

I think `jotai-suspense` is a good one. (That's something we did for `atomWithPending`.)
Would anyone like to open a PR in https://github.com/jotaijs/jotai-suspense ?

**Reply by HarunKilic** [reply]
It would be awesome to have!
